cache.go

```diff
@@ -1,91 +1,266 @@
+// LRU functionality largely inspired by https://github.com/golang/groupcache
 package agecache
 
 import (
+	"container/list"
+	"errors"
 	"sync"
 	"sync/atomic"
 	"time"
-
-	"github.com/hashicorp/golang-lru/simplelru"
 )
 
 // Stats hold cache statistics.
 type Stats struct {
-	Hits   int64 // number of cache hits.
-	Misses int64 // number of cache misses
+	Count     int64 // Gauge, number of items in the cache
+	Sets      int64 // Counter, number of sets
+	Gets      int64 // Counter, number of gets
+	Hits      int64 // Counter, number of cache hits
+	Misses    int64 // Counter, number of cache misses
+	Evictions int64 // Counter, number of evictions
 }
 
-// Sub subtracts stats `s` with `o`.
-func (s Stats) Sub(o Stats) Stats {
+// Delta returns a Stats object such that all counters are calculated as the
+// difference since the previous.
+func (stats Stats) Delta(previous Stats) Stats {
 	return Stats{
-		Hits:   atomic.LoadInt64(&s.Hits) - o.Hits,
-		Misses: atomic.LoadInt64(&s.Misses) - o.Misses,
+		Count:     stats.Count,
+		Sets:      stats.Sets - previous.Sets,
+		Gets:      stats.Gets - previous.Gets,
+		Hits:      stats.Hits - previous.Hits,
+		Misses:    stats.Misses - previous.Misses,
+		Evictions: stats.Evictions - previous.Evictions,
 	}
 }
 
-type entry struct {
+// Configuration for the Cache.
+type Config struct {
+	// Maximum number of items in the cache
+	Size int
+	// Optional time before an item expires
+	MaxAge time.Duration
+	// Optional callback invoked when an item is evicted due to the LRU policy
+	OnEviction func(key, value interface{})
+	// Optional callback invoked when an item expired
+	OnExpiration func(key, value interface{})
+}
+
+// Entry pointed to by each list.Element
+type cacheEntry struct {
+	key     interface{}
 	value   interface{}
 	created time.Time
 }
 
 // LRU implements a thread-safe fixed size LRU cache.
 type Cache struct {
-	lru    *simplelru.LRU
-	mu     sync.Mutex
-	maxAge time.Duration
-	stats  Stats
-}
-
-// New constructs an LRU of the given size and max age to return
-// results for
-func New(size int, maxAge time.Duration) (*Cache, error) {
-	l, err := simplelru.NewLRU(size, nil)
-	if err != nil {
-		return nil, err
+	// Fields defined by configuration
+	size         int
+	maxAge       time.Duration
+	onEviction   func(key, value interface{})
+	onExpiration func(key, value interface{})
+
+	// Cache statistics
```

Comment:
curios, why set those on the `*Cache` struct and not have a value of `stats` around?

Reply:
If we pass stats around, then either we:
a) Update stats.Capacity whenever evictionList.Len() changes.
b) Are ok with the fact that at any given point, the stats object is outdated and needs to have its Count field updated before returning it to the user. Seems messy

* Note that evictionList.Len() is a O(1) op

Reply:
gotcha, makes sense!